python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -0,0 +1,575 @@
+        return result
+
+    def _generateGrid(self, exposure, forceEvenSized=False):
+        """Generate two lists of bounding boxes that evenly grid `exposure`
+
+        Unless the config was provided with `centroidCoordsX` and
+        `centroidCoordsY`, grid (subimage) centers are spaced
+        evenly by gridStepX/Y. Then the grid is adjusted as
+        little as possible to evenly cover the input exposure (if
+        adjustGridOption is not 'none'). Then the second set of
+        bounding boxes is expanded by borderSizeX/Y. The expanded
+        bounding boxes are adjusted to ensure that they intersect
+        the exposure's bounding box. The resulting lists of bounding
+        boxes and corresponding expanded bounding boxes are
+        returned, and also set to `self.boxes0`, `self.boxes1`.
+
+        Parameters
+        ----------
+        exposure : afwImage.Exposure`
+            input exposure whose full bounding box is to be evenly gridded.
+        forceEvenSized : boolean
+            force grid elements to have even-valued x- and y- dimensions? (for ZOGY)
+
+        Returns
+        -------
+        tuple containing two lists of `afwGeom.BoundingBox`es
+        """
+        # Extract the config parameters for conciseness.
+        gridSizeX = self.config.gridSizeX
+        gridSizeY = self.config.gridSizeY
+        gridStepX = self.config.gridStepX
+        gridStepY = self.config.gridStepY
+        borderSizeX = self.config.borderSizeX
+        borderSizeY = self.config.borderSizeY
+        adjustGridOption = self.config.adjustGridOption
+        scaleByFwhm = self.config.scaleByFwhm
+        bbox = exposure.getBBox()
+
+        psfFwhm = (exposure.getPsf().computeShape().getDeterminantRadius() *
+                   2. * np.sqrt(2. * np.log(2.)))
+
+        def rescaleValue(val):
+            if scaleByFwhm:
+                return np.rint(val * psfFwhm).astype(int)
+            else:
+                return np.rint(val).astype(int)
+
+        gridSizeX = rescaleValue(gridSizeX)
+        if gridSizeX > bbox.getWidth():
+            gridSizeX = bbox.getWidth()
+        gridSizeY = rescaleValue(gridSizeY)
+        if gridSizeY > bbox.getHeight():
+            gridSizeY = bbox.getHeight()
+        gridStepX = rescaleValue(gridStepX)
+        if gridStepX > bbox.getWidth():
+            gridStepX = bbox.getWidth()
+        gridStepY = rescaleValue(gridStepY)
+        if gridStepY > bbox.getHeight():
+            gridStepY = bbox.getHeight()
+        borderSizeX = rescaleValue(borderSizeX)
+        borderSizeY = rescaleValue(borderSizeY)
+
+        nGridX = bbox.getWidth() // gridStepX
+        nGridY = bbox.getHeight() // gridStepY
+
+        if adjustGridOption == 'spacing':
+            nGridX = bbox.getWidth() / gridStepX
+            # Readjust gridStepX so that it fits perfectly in the image.
+            gridStepX = float(bbox.getWidth() - gridSizeX) / float(nGridX)
+            if gridStepX <= 0:
```

Comment:
This is just to catch a case which shouldn't happen. I'll raise an exception instead.